tools/Google.Cloud.Tools.Analyzers/PublicDependencyForbiddenAnalyzer.cs

```diff
@@ -0,0 +1,280 @@
+                       methodSymbol.Parameters[i].Type,
+                       methodNode => methodNode.ParameterList.Parameters[i].Type.GetLocation());
+                }
+            }
+        }
+
+        private static void AnalyzeNamedType(SymbolAnalysisContext context)
+        {
+            var typeSymbol = (INamedTypeSymbol)context.Symbol;
+            CheckType<TypeDeclarationSyntax>(
+                context,
+                typeSymbol,
+                methodNode => methodNode.BaseList.GetLocation());
+        }
+
+        private static void AnalyzeProperty(SymbolAnalysisContext context)
+        {
+            var propertySymbol = (IPropertySymbol)context.Symbol;
+            CheckType<PropertyDeclarationSyntax>(
+                context,
+                propertySymbol.Type,
+                propertyNode => propertyNode.Type.GetLocation());
+
+            for (int i = 0; i < propertySymbol.Parameters.Length; ++i)
+            {
+                CheckType<IndexerDeclarationSyntax>(
+                    context,
+                    propertySymbol.Parameters[i].Type,
+                    propertyNode => propertyNode.ParameterList.Parameters[i].Type.GetLocation());
+            }
+        }
+
+        private static void CheckType<T>(SymbolAnalysisContext context, ITypeSymbol type, Func<T, Location> getErrorLocation)
+            where T : CSharpSyntaxNode
+        {
+            if (!context.Symbol.IsExternallyVisible())
+            {
+                return;
+            }
+
+            if (ForbiddenNamespacePrefixes.Any(context.Symbol.ContainingSymbol.ToDisplayString().StartsWith))
+            {
+                // If a symbol is actually defined in a forbidden namespace, don't add an error to it.
+                return;
+            }
+
+            var checkedTypes = new Dictionary<ITypeSymbol, TypeCheckResult>();
+            var result = CheckType(type, checkedTypes);
+            if (!result.IsTypeForbidden)
+            {
+                return;
+            }
+
+            if (context.Symbol.DeclaringSyntaxReferences.IsEmpty)
+            {
+                return;
+            }
+
+            T declaringNode = (T)context.Symbol.DeclaringSyntaxReferences.FirstOrDefault().GetSyntax(context.CancellationToken);
+            context.ReportDiagnostic(
+                Diagnostic.Create(
+                    Rule,
+                    getErrorLocation(declaringNode),
+                    type.ToDisplayString(),
+                    result.ForbiddenNamespace));
+        }
+
+        private static TypeCheckResult CheckType(ITypeSymbol type, Dictionary<ITypeSymbol, TypeCheckResult> checkedTypes)
+        {
+            if (!checkedTypes.TryGetValue(type, out var result))
+            {
+                // Add a placeholder before we do anything else to prevent recursion issues.
+                checkedTypes.Add(type, TypeCheckResult.Allowed);
+
+                result = CheckTypeImpl(type);
```

Comment:
This is a really interesting use for local methods (do something with the return value in every case, even if there are multiple return statements). May put this into C# in Depth :)

Reply:
Thanks. I'm not sure if it's the cleanest way to do it, but I like how it let me memoize some functionality without exposing the raw functionality on the class and without having to introduce a whole bunch of extra if/elses.